fpga_interchange/pseudo_pip_model.h

```diff
@@ -0,0 +1,108 @@
+/*
+ *  nextpnr -- Next Generation Place and Route
+ *
+ *  Copyright (C) 2021  Symbiflow Authors
+ *
+ *
+ *  Permission to use, copy, modify, and/or distribute this software for any
+ *  purpose with or without fee is hereby granted, provided that the above
+ *  copyright notice and this permission notice appear in all copies.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ *  WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ *  MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ *  ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ *  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ *  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ *  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ */
+
+#ifndef PSEUDO_PIP_MODEL_H
+#define PSEUDO_PIP_MODEL_H
+
+#include <tuple>
+
+#include "nextpnr_namespaces.h"
+#include "nextpnr_types.h"
+#include "site_router.h"
+#include "dynamic_bitarray.h"
+#include "hash_table.h"
+
+NEXTPNR_NAMESPACE_BEGIN
+
+struct PseudoPipBel {
+    // Which BEL in the tile does the pseudo pip use?
+    int32_t bel_index;
+
+    // What is the index of the input BEL pin that the pseudo pip used?
+    //
+    // NOTE: This is **not** the name of the pin.
+    int32_t input_bel_pin;
+
+    // What is the index of the output BEL pin that the pseudo pip used?
+    //
+    // NOTE: This is **not** the name of the pin.
+    int32_t output_bel_pin;
+};
+
+// Storage for tile type generic pseudo pip data and lookup.
+struct PseudoPipData {
+    // Initial data for specified tile type, if not already initialized.
+    void init_tile_type(const Context *ctx, int32_t tile_type);
+
+    // Get the highest PipId::index found in a specified tile type.
+    size_t get_max_pseudo_pip(int32_t tile_type) const;
+
+    // Get the list of possible sites that a pseudo pip might be used in.
+    const std::vector<size_t> &get_possible_sites_for_pip(const Context *ctx, PipId pip) const;
+
+    // Get list of BELs the pseudo pip uses, and how it routes through them.
+    //
+    // This does **not** include site ports or site pips.
+    const std::vector<PseudoPipBel> &get_logic_bels_for_pip(const Context *ctx, int32_t site, PipId pip) const;
+
+    HashTables::HashMap<int32_t, size_t> max_pseudo_pip_for_tile_type;
+    HashTables::HashMap<std::pair<int32_t, int32_t>, std::vector<size_t>> possibles_sites_for_pip;
+    HashTables::HashMap<std::tuple<int32_t, int32_t, int32_t>, std::vector<PseudoPipBel>> logic_bels_for_pip;
```

Comment:
by three elements I think a custom key type is nicer than a tuple

Reply:
Done.